holmes/plugins/toolsets/logging_api.py

```diff
@@ -0,0 +1,205 @@
+    StructuredToolResult,
+    Tool,
+    ToolParameter,
+    Toolset,
+)
+from holmes.plugins.toolsets.utils import get_param_or_raise
+
+# Default values for log fetching
+DEFAULT_LOG_LIMIT = 2000
+DEFAULT_TIME_SPAN_SECONDS = 3600
+
+
+class LoggingLabelsConfig(BaseModel):
+    """Common label mapping configuration for all logging backends"""
+
+    pod: Optional[str] = None
+    namespace: Optional[str] = None
+
+
+class LoggingConfig(BaseModel):
+    """Base configuration for all logging backends"""
+
+    labels: Optional[LoggingLabelsConfig] = None
+
+
+class FetchLogsParams(BaseModel):
+    namespace: str
+    pod_name: str
+    start_time: Optional[str] = None
+    end_time: Optional[str] = None
+    filter_pattern: Optional[str] = None
+    limit: Optional[int] = None
+
+
+class BaseLoggingToolset(Toolset, ABC):
+    """Base class for all logging toolsets"""
+
+    @abstractmethod
+    def fetch_logs(self, params: FetchLogsParams) -> StructuredToolResult:
+        pass
+
+
+class LoggingTool(Tool):
+    """Common tool for fetching pod logs across different logging backends"""
+
+    def __init__(self, toolset: BaseLoggingToolset):
+        super().__init__(
+            name="fetch_logs",
+            description="Fetch logs for a Kubernetes pod",
+            parameters={
+                "pod_name": ToolParameter(
+                    description="The exact kubernetes pod name",
+                    type="string",
+                    required=True,
+                ),
+                "namespace": ToolParameter(
+                    description="Kubernetes namespace", type="string", required=True
+                ),
+                "start_time": ToolParameter(
+                    description="Start time for logs. Can be an RFC3339 formatted timestamp (e.g. '2023-03-01T10:30:00Z') for absolute time or a negative integer (e.g. -3600) for relative seconds before end_time.",
+                    type="string",
+                    required=False,
+                ),
+                "end_time": ToolParameter(
+                    description="End time for logs. Must be an RFC3339 formatted timestamp (e.g. '2023-03-01T12:30:00Z'). If not specified, defaults to current time.",
+                    type="string",
+                    required=False,
+                ),
+                "limit": ToolParameter(
+                    description="Maximum number of logs to return",
+                    type="integer",
+                    required=False,
+                ),
+                "filter": ToolParameter(
+                    description="Optional filter to apply to logs (regex pattern)",
+                    type="string",
+                    required=False,
+                ),
+            },
+        )
+        self._toolset = toolset
+
+    def _invoke(self, params: dict) -> StructuredToolResult:
+        structured_params = FetchLogsParams(
+            namespace=get_param_or_raise(params, "namespace"),
```

Comment:
Do we really need get_param_or_raise? wouldn't it fail anyway on the pydantic validation of the field?

Reply:
pydantic validation generates atrocious validation messages.
Linting will fail if the param is passed as is.

Reply:
What im concern with the current implementation is that we need to remember to add the get_param_or_raise for all required params from now on. To be honest it is something that we will forget.

So either we need to have some function that checks the models and validate it or add logic that catch pydantic ValidationError and handle it as we expected. You can see example here: https://docs.pydantic.dev/latest/errors/errors/#customize-error-messages

Reply:
Yes, agree, and this can be part of a toolset refactor. Added to https://robusta-dev.atlassian.net/browse/ROB-1243